### DISCUSSION TITLE
defaut-directory changes to inside .git/ subdirectory in COMMIT_MSG buffer

### DISCUSSION THREAD

**Discussion by apteryks** [magit/magit#5271] [Support] [answered]
Hello!

Please explain

    (1) what behavior you expected

When pressing c twice (`c c`) in the `magit: $project` buffer to author a commit, the current directory (`default-directory` variable in Emacs) should be set to the project top directory.

    (2) what behavior you observed

Instead, it gets set to the `.git/ subdirectory`, or a `.git/worktrees/$worktree` subdirectory.  This breaks diff-operating commands such as `C` (`magit-add-log`), and I need to do `M-x cd $correct_dir` and C-c C-d to refresh the diff on each commit, which gets annoying quickly.

    (3) and how we can reproduce the issue.

I'm using Emacs 29.4 with Magit 4.1.3.  I can reproduce this on my Guix checkout, starting Emacs wih `emacs -q`, manually loading the Magit library with `M-x load magit` then attempting to author a commit in that checkout.

Post the output of:

    M-x magit-version RET

```
Magit 4.1.3, Transient 0.8.1, Git 2.46.0, Emacs 29.4, gnu/linux
```
[...]
Alternatively, run:

    M-x magit-emacs-Q-command RET

I can reproduce the problem in such an environment; the above command returned:
```
/home/maxim/.guix-profile/bin/emacs -Q --eval \(setq\ debug-on-error\ t\) -L /gnu/store/5fz4hfsqgh2rgdz2gxpxsnf6ak2nbsnx-emacs-compat-30.0.0.0/share/emacs/site-lisp/compat-30.0.0.0/ -L /gnu/store/225hahh22k3bsxxlq479463lwsxj8fix-emacs-dash-2.19.1/share/emacs/site-lisp/dash-2.19.1/ -L /gnu/store/cxsacyan78s9p1l1fnkri5q370n30ngy-emacs-transient-0.8.1/share/emacs/site-lisp/transient-0.8.1/ -L /gnu/store/yb845lyz9d7ycidrdzc8qrkj9bhy4vyr-emacs-with-editor-3.4.3/share/emacs/site-lisp/with-editor-3.4.3/ -L /gnu/store/mn0hmr3rk52zgqh8jwml8cfyg6hjv7ia-emacs-magit-4.1.3/share/emacs/site-lisp/magit-4.1.3/ -l /gnu/store/mn0hmr3rk52zgqh8jwml8cfyg6hjv7ia-emacs-magit-4.1.3/share/emacs/site-lisp/magit-4.1.3/magit
```

Thank you!

**Comment by apteryks** [comment]
It doesn't seem to be a regression caused by git, as I've tried reverting from git 2.46 to 2.45.2 and even 2.41.0, but all reproduce the problem.

**Comment by apteryks** [comment]
Hm, but maybe it'd work, as it seems it was always the case that the default-directory when editing the COMMIT_MSG file would be the directory it's created in, which is inside the .git/ directory (c.f.: https://github.com/magit/magit/issues/5100#issuecomment-1967390767). I'll retry with 2.41.0, trying `magit-add-log` instead of just looking at the value of the `default-directory` variable.

**Comment by apteryks** [comment]
Yes, it works with git 2.41.0 even when the `COMMIT_EDITMSG` and `magit-diff: .git` buffers' `default-directory` is the `.git/` project subdirectory.

**Comment by apteryks** [comment]
It also works using git 2.45.2... and... in 2.46.0 !??.

I must be nuts.  I swear this was not working, even after restarting my emacs server.  I've since rebooted, though...  I'll close it for now, and will reopen if I manage to make sense of how to reproduce it.

**Comment by tarsius** [comment, marked as answer]
This is the expected default behavior but there is an option to control it. Its docstring also explains why the default is the default:

```
git-commit-cd-to-toplevel is a variable defined in `git-commit.el'.

Its value is nil

Whether to set `default-directory' to the worktree in message buffer.

Editing a commit message is done by visiting a file located in the git
directory, usually "COMMIT_EDITMSG".  As is done when visiting any
file, the local value of `default-directory' is set to the directory
that contains the file.

If this option is non-nil, then the local `default-directory' is changed
to the working tree from which the commit command was invoked.  You may
wish to do that, to make it easier to open a file that is located in the
working tree, directly from the commit message buffer.

If the git variable `safe.bareRepository' is set to "explicit", then
you have to enable this, to be able to commit at all.  See issue #5100.

This option only has an effect if the commit was initiated from Magit.
```

> This breaks diff-operating commands such as `C` (`magit-add-log`),

> This breaks diff-operating commands such as C (magit-add-log), and I need to do M-x cd $correct_dir and C-c C-d to refresh the diff on each commit, which gets annoying quickly.

I suppose you mean `magit-commit-add-log`. That command *is supposed* to be run from a diff -- how else would it know what the "current change" is?

**Reply by apteryks** [reply]
Yes, I wasn't precise enough; my problem was that `magit-commit-add-log`, when executed on some diff hunk (on a magit diff buffer) would fail, with some 'not a git repository' error or similar.  Thanks for mentioning `git-commit-cd-to-toplevel`; it might come in handy.

**Reply by apteryks** [reply]
Another thing I've noticed; if the `default-directory` in the `COMMIT_EDITMSG` buffer is e.g. `/home/maxim/src/guix/.git/worktrees/guix-gnome-team/`, if I try to find a file with `counsel-git`, it finds a file in the main tree (`~/src/guix`), not in my worktree (`~/src/guix-gnome-team`).  Would `git-commit-cd-to-toplevel` help with this?

**Reply by apteryks** [reply]
I just tried to set `git-commit-cd-to-toplevel` via `customize-option`; it works well for the above use case. Thank you :-).

**Reply by apteryks** [reply]
I've just had another `default-directory` set to `~/src/guix/.git/`, which apparently breaks navigating to the latest file touched by a diff hunk (`C-RET` on the diff in the `magit-diff: .git` buffer.  It also causes this kind of message to appear when using `magit-commit-add-log`, visible in the `*Messages*` buffer: 
```
Use M-x make-directory RET RET to create the directory and its parents
process-file: Setting current directory: Aucun fichier ou dossier de ce nom, /home/maxim/src/guix/.git/gnu/packages/
```
The `git-commit-cd-to-toplevel` is set to `t`.

**Reply by apteryks** [reply]
> I've just had another `default-directory` set to `~/src/guix/.git/`, which apparently breaks navigating to the latest file touched by a diff hunk (`C-RET` on the diff in the `magit-diff: .git` buffer. It also causes this kind of message to appear when using `magit-commit-add-log`, visible in the `*Messages*` buffer:
> 
> ```
> Use M-x make-directory RET RET to create the directory and its parents
> process-file: Setting current directory: Aucun fichier ou dossier de ce nom, /home/maxim/src/guix/.git/gnu/packages/
> ```
> 
> The `git-commit-cd-to-toplevel` is set to `t`.

I finally fully investigated this, and it was the `.git/config` file lacking the `bare = false` line in the `[core]` section, which was causing the `git rev-parse --is-bare-repository` to return `true` instead of `false`, which is used by `magit-toplevel`.  Phew!
